### ISSUE TITLE
Meta-SGD - Learning rate alpha not changing

### ISSUE THREAD

**Issue by janbolle** [learnables/learn2learn#167]
1) I have a question regarding the Meta-SGD implementation. Am I right, that the inner-learning rate \alpha will be adapted?
How is this done in the implementation? I really don’t get it.
Also if I output the learning rate at the end of the outer loop (`print(‘lrs’, meta_learner.lrs.state_dict()[‘0’])`)
I don’t see changes…

2) I added:
```python
        opt.zero_grad()  # necessary?

        second_order = False
        from torch.autograd import grad
        gradients = grad(adaptation_loss,
                         meta_learner.lrs,  # get new gradients for adaption of \alpha (lrs)
                         retain_graph=second_order,
                         create_graph=second_order)

        for p, g in zip(meta_learner.lrs, gradients):
            with torch.no_grad():
                p -= meta_lr * g    # update every parameter of \alpha (lrs) with \beta (meta_lr)
        print('lrs', meta_learner.lrs.state_dict()['0'])
```
at the end of metasgd_a2c.py.
Now I can see \alpha adapt. Is this correct?

3) Also, I don't know if this is the most efficient way in torch to do this...

Any thoughts?

**Comment by seba-1511** [comment]
Hello @janbolle,

With Meta-SGD, the alphas are not adapted during fast-adaptation. However, they should be learned and updated after each call to `opt.step()`. 

If you’re looking to adapt the alphas during fast-adaptation, check out Alpha-MAML. It’s recently supported by `learn2learn`, and you can implement it by replacing `MetaCurvatureTransform` with `ModuleTransform(l2l.nn.Scale)` in the [Meta-Curvature example](https://github.com/learnables/learn2learn/blob/master/examples/vision/metacurvature_fc100.py). (Also make sure to pass `adapt_transform=True` to `GBML`.)

**Comment by janbolle** [comment]
@seba-1511 Thanks for the reply.
I want to update \alpha in the outer-loop but as far as I see it, this is not the case in the Meta-SGD example...

In that [gist](https://gist.github.com/janbolle/b4557b4b552b6b45b207c108050daa24), if you set `outer_alpha=False` you can see, that \alpha is never changing.
So I assume `opt.step()`does not update \alpha.

I tried to implement the update of \alpha, which you can enable setting `outer_alpha=True`

**Comment by seba-1511** [comment]
Hi @janbolle ,

From your gist I think the simple fix is to change [line 83](https://gist.github.com/janbolle/b4557b4b552b6b45b207c108050daa24#file-metasgd_a2c-py-L83) from `optim.Adam(policy.parameters(), lr=meta_lr)` to `optim.Adam(meta_learner.parameters(), lr=meta_lr)`. The alphas are stored within the meta_learner module, so the optimizer cannot update them when we `opt.step()`.

**Comment by janbolle** [comment]
@seba-1511 , yes, this fixed it. \alpha is now changing, thank you for your fast reply!
